client/blockchain-service/src/utils.rs

```diff
@@ -2082,4 +2079,333 @@ where
+                        }
+                        Some(ManagedProvider::Msp(_)) => {
+                            self.msp_process_finality_events(block_hash, ev.event.clone().into());
+                        }
+                        _ => {}
+                    }
+                }
+            }
+            Err(e) => {
+                // TODO: This can happen for older blocks where state has been pruned, or if
+                // we're parsing a block authored with an older version of the runtime
+                // using a node that has a newer version of the runtime. Consider using runtime APIs
+                // for getting old data of previous blocks, and this just for current blocks.
+                error!(
+                        target: LOG_TARGET,
+                        "Failed to get events for block {:?}: {:?}",
+                        block_hash, e
+                );
+            }
+        }
+    }
+
+    /// Catch up on any blocks that were imported to the database but not processed.
+    ///
+    /// This handles race conditions during shutdown where the Substrate client database commits a block write
+    /// before our handlers fully processed it. On restart, we compare `last_processed_block`
+    /// (from persistent storage) with the database's `best_block` and process any gap.
+    ///
+    /// This function also initializes `self.best_block` to the client's actual best block,
+    /// ensuring we start from the correct position regardless of whether there are missed blocks.
+    pub(crate) async fn catch_up_missed_blocks(&mut self) {
+        // Get the best block saved in the node's Substrate client database
+        let chain_info = self.client.info();
+        let best_number: BlockNumber<Runtime> = chain_info.best_number.saturated_into();
+        let best_hash = chain_info.best_hash;
+
+        // Initialize self.best_block to match the client's actual state
+        self.best_block = MinimalBlockInfo {
+            number: best_number,
+            hash: best_hash,
+        };
+
+        // Get the last processed block saved in the node's persistent storage
+        let Some(last_processed_number) = self.get_last_processed_block() else {
+            // If there's no `last_processed_block` in persistent storage, it means this is the first startup of the node, which
+            // means there's nothing to catch up on.
+            info!(target: LOG_TARGET, "No last processed block found in persistent storage. Skipping startup catch up.");
+            return;
+        };
+
+        // If the best block is not greater than the last processed block, there's nothing to catch up on
+        if best_number <= last_processed_number {
+            info!(
+                target: LOG_TARGET,
+                "☑️ No missed blocks to catch up (last_processed={}, best={})",
+                last_processed_number, best_number
+            );
+            return;
+        }
```

Comment:
Question regarding this check. While the node is down, there could have been a reorg and let's say the `last_processed_number` is still greater or equal to `best_number`. Wouldn't we have to check that there wasn't a reorg based on the tree route as well?

But I guess this will get caught anyways by the `block_import_notification_stream` anyways.